OCPBUGS-46493: Raise the minSyncPeriod for NetworkPolicies to 10s during migration

danwinship opened the pull request:
Normally in SDN, NetworkPolicies that match an entire namespace are implemented by matching the VNID, but this doesn't work during sdn->ovn migration, so they get rewritten to use pod IPs instead. The code for doing this is not really optimized, and particularly during migration (when every pod in the cluster gets deleted from its original node and then recreated on another) it results in the NP code doing a huge amount of work (and, seemingly, generating new garbage faster than the garbage collector can collect it).

This tries to mitigate the problem by only doing resyncs once every 10 seconds in migration mode (as opposed to once a second in non-migration mode). Hopefully this is enough to get the existing customers through migration. (If this does work I have other ideas, but they require rewriting a large chunk of code.)

(See also #650 for a more complicated, but maybe better, approach.)

/assign @pliurh

danwinship committed:
Raise the minSyncPeriod for NetworkPolicies during migration in large clusters

The hack to make ovn->sdn traffic still obey NetworkPolicies results
in us having to do A LOT more work in large clusters. Try to mitigate
the problem by just doing it less often...

danwinship commented:
/jira refresh

danwinship committed:
Fix handling of default namespace in migration mode

The default namespace is equivalent to the host-network namespace in
openshift-sdn.

danwinship commented:
/test unit

npinaeva commented on pkg/network/node/networkpolicy.go:
```diff
@@ -189,10 +189,16 @@ func (np *networkPolicyPlugin) Start(node *OsdnNode) error {
+	if np.inMigrationMode {
+		// The existing code can't keep up with the churn rate in large clusters
+		// when in migration mode.
+		minSyncPeriod = 10 * time.Second
```
is there a chance of wrongly allowing or denying existing connections for 10s because of this change?

danwinship replied:
So of course, in the affected clusters, this is basically already happening because of slow syncs.

If someone modifies the NetworkPolicies or Pod/Namespace labels, then it will take us up to 10s to catch up to that modification. But there is no reason why someone would need to do that in the middle of a migration, so this doesn't seem like a huge problem.

It does mean that pods will be fully-isolated for up to 10s on startup. It would be good to avoid that, but the goal of this particular PR was to be simple... (I couldn't even come up with a good way to make the behavior be conditional on "it's a big cluster", because this code runs before the informer caches have fully synced, and there's no way to change the parameters of the BoundedFrequencyRunner after creating it.)

npinaeva replied:
My concern is that smaller clusters that didn't have this extra delay will now get it, trying to figure out what the result of this could be.
Pods that are migrating to the new node will be ignored for 10s I guess, so if a migrated pod tries to immediately reach a non-migrated pod after its startup, this could fail?

danwinship replied:
yes, though presumably it will retry. But you'll get additional delay, and spurious error messages in your pod logs.

danwinship committed:
Copy TestNetworkPolicy to _TestNetworkPolicyInMigrationMode

We want the migration mode test to be as comprehensive as the
non-migration-mode test, so start out by making a copy of the
original. (Ideally we'd have both tests use a single helper function,
but the differences between them are large enough to make that
annoying.)

In this commit, the new migration mode test is disabled (initial "_"
in its name) because the test assertions haven't been updated.

danwinship committed:
Fix up new TestNetworkPolicyInMigrationMode, remove old one

danwinship committed:
Add another test to TestNetworkPolicy*

Ensure that the right namespaces are recalculated when pods change.

danwinship commented:
OK, we tried to test this with random minSyncPeriod, but it's hard to test because only _some_ of the pods in any run will have the memory usage problems (and it takes a _really_ long time to do a test run). So rather than delay this further, we're just going to stick with 10s, which we know works. (Though in the latest version, the 10s now only applies to clusters with at least 100 nodes; smaller clusters still use the old value.)

danwinship commented:
/label backport-risk-assessed

npinaeva commented:
/lgtm

asood-rh commented:
/label cherry-pick-approved
